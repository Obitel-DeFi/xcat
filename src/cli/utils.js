import chalk from 'chalk'
import {existsSync, readFileSync} from 'fs'

import Trade from '../trade'
import Config from '../config'

const logError = msg => console.error(chalk.red(`\nERROR: ${msg}`))

const fileToObj = filename => JSON.parse(readFileSync(filename).toString())

const verifyArgTradeFile = filename => {
  if (!filename || (typeof filename === 'string' && filename.trim() === '')) {
    logError(`No file provided. Pass the name of a trade.json file.`)
    return false
  }

  if (!existsSync(filename)) {
    logError(`File [${filename}] does not exist.`)
    return false
  }

  try {
    new Trade(fileToObj(filename))
  } catch (err) {
    logError(`Failed to parse [${filename}]:`)
    console.error(`\n\n${err}`)
    return false
  }

  return true
}

const verifyConfigFile = filename => {
  // look for config.json in local directory if not provided
  if (!filename || typeof filename !== 'string' || filename.trim() === '')
    filename = './config.json'

  if (!existsSync(filename)) {
    logError(
      `File [${filename}] does not exist. Put a config.json in the ` +
        `local path OR provide one with --config.`
    )
    return false
  }

  try {
    new Config(fileToObj(filename))
  } catch (err) {
    logError(`Failed to parse [${filename}]:`)
    console.error(`\n\n${err}`)
    return false
  }

  return true
}

const verifyArgTradeID = id => {
  throw new Error(
    'not yet implemented - will pull a trade record from a local database'
  )
}

export {
  fileToObj,
  logError,
  verifyArgTradeFile,
  verifyArgTradeID,
  verifyConfigFile,
}
