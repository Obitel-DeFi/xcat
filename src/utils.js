import crypto from 'crypto'
import {URL} from 'url'
import stellarSdk from 'stellar-sdk'
import Web3Utils from 'web3-utils'
import hasIn from 'lodash/hasIn'

// Format required for sending bytes through an eth js client:
//  - hex string representation
//  - prefixed with 0x
const bufToStr = b => '0x' + b.toString('hex')

const clone = obj => JSON.parse(JSON.stringify(obj))

const sha256 = x =>
  crypto
    .createHash('sha256')
    .update(x)
    .digest()

const random32 = () => crypto.randomBytes(32)

const newSecretHashPair = () => {
  const secret = random32()
  const hash = sha256(secret)
  return {
    secret: bufToStr(secret),
    hash: bufToStr(hash),
  }
}

const isStellarPublicAddress = addr =>
  stellarSdk.StrKey.isValidEd25519PublicKey(addr)
const isStellarSecretSeed = addr =>
  stellarSdk.StrKey.isValidEd25519SecretSeed(addr)
const isStellarFederatedAddress = addr => /^[^*,]*\*[a-z0-9-.]*$/i.test(addr)

const isEthereumPublicAddress = addr => Web3Utils.isAddress(addr)

const isUrl = url => {
  try {
    new URL(url)
    return true
  } catch (e) {
    return false
  }
}

const isClassWithName = (cls, name) =>
  hasIn(cls, 'constructor.name') && cls.constructor.name === name

export {
  bufToStr,
  clone,
  newSecretHashPair,
  random32,
  sha256,
  isStellarPublicAddress,
  isStellarSecretSeed,
  isStellarFederatedAddress,
  isEthereumPublicAddress,
  isClassWithName,
  isUrl,
}
