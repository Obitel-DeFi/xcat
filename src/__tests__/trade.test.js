import expect from 'expect'
import {clone} from '../utils'
import Trade from '../trade'

const trade1 = require('./data/trade1.json')

describe('trade', () => {
  it('creates new Trade that has all properties of the given trade object', () => {
    const trade = new Trade(trade1)
    Object.keys(trade1).forEach(attr =>
      expect(trade[attr]).toEqual(trade[attr])
    )
  })

  it('throws if trade has a bad stellar address', () => {
    const tradeBadAddr = clone(trade1)
    tradeBadAddr.stellar.depositor = 'GDDADDDA'
    expect(() => new Trade(tradeBadAddr)).toThrowErrorMatchingSnapshot()
  })

  it('throws if trade has a bad ethereum address', () => {
    const tradeBadAddr = clone(trade1)
    tradeBadAddr.ethereum.withdrawer = '0xabcdef'
    expect(() => new Trade(tradeBadAddr)).toThrowErrorMatchingSnapshot()
  })
})
