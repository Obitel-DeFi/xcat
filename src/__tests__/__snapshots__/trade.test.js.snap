// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`trade throws if trade has a bad ethereum address 1`] = `
"trade doesn't conform to schema trade.json [
	message: \\"data.ethereum.withdrawer should match format \\"ethereumPublicAddress\\"\\"
	raw: [
  {
    \\"keyword\\": \\"format\\",
    \\"dataPath\\": \\".ethereum.withdrawer\\",
    \\"schemaPath\\": \\"#/properties/ethereum/properties/withdrawer/format\\",
    \\"params\\": {
      \\"format\\": \\"ethereumPublicAddress\\"
    },
    \\"message\\": \\"should match format \\\\\\"ethereumPublicAddress\\\\\\"\\"
  }
]]"
`;

exports[`trade throws if trade has a bad stellar address 1`] = `
"trade doesn't conform to schema trade.json [
	message: \\"data.stellar.depositor should match format \\"stellarPublicAddress\\", data.stellar.depositor should match format \\"stellarFederatedAddress\\", data.stellar.depositor should match exactly one schema in oneOf\\"
	raw: [
  {
    \\"keyword\\": \\"format\\",
    \\"dataPath\\": \\".stellar.depositor\\",
    \\"schemaPath\\": \\"#/properties/stellar/properties/depositor/oneOf/0/format\\",
    \\"params\\": {
      \\"format\\": \\"stellarPublicAddress\\"
    },
    \\"message\\": \\"should match format \\\\\\"stellarPublicAddress\\\\\\"\\"
  },
  {
    \\"keyword\\": \\"format\\",
    \\"dataPath\\": \\".stellar.depositor\\",
    \\"schemaPath\\": \\"#/properties/stellar/properties/depositor/oneOf/1/format\\",
    \\"params\\": {
      \\"format\\": \\"stellarFederatedAddress\\"
    },
    \\"message\\": \\"should match format \\\\\\"stellarFederatedAddress\\\\\\"\\"
  },
  {
    \\"keyword\\": \\"oneOf\\",
    \\"dataPath\\": \\".stellar.depositor\\",
    \\"schemaPath\\": \\"#/properties/stellar/properties/depositor/oneOf\\",
    \\"params\\": {},
    \\"message\\": \\"should match exactly one schema in oneOf\\"
  }
]]"
`;
